### PR TITLE
[RW-5045][risk=no] Hail Microarray Demo codegen

### PR DESCRIPTION
**Pull request by ericsong**
Description:

Example of generated code

image

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI + API server with `yarn test-local`
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

**Review by als364 — APPROVED** [review]

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java [hunk 1]
```diff
@@ -815,6 +815,58 @@ private String getQualifiedColumnName(Domain currentDomain, String columnName) {
             + "head results.P2.assoc");
   }
 
+  @Override
+  public List<String> generateHailDemoCode(String qualifier) {
+    final String phenotypeFilename = "phenotypes_annotations_" + qualifier + ".txt";
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:820]: .tsv?

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java [hunk 2]
```diff
@@ -815,6 +815,58 @@ private String getQualifiedColumnName(Domain currentDomain, String columnName) {
...
+    final String cohortQualifier = "cohort_" + qualifier;
+    final String cohortVcfFilename = cohortQualifier + ".vcf";
+    final String cohortMtFilename = cohortQualifier + ".mt";
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:823]: Please spell it out in the variable name

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java [hunk 3]
```diff
@@ -815,6 +815,58 @@ private String getQualifiedColumnName(Domain currentDomain, String columnName) {
...
+
+    return ImmutableList.of(
+        "import subprocess, os\n"
+            + "import random\n"
+            + "\n"
+            + "# Creating phenotype annotations file\n"
+            + "phenotypes_table = []\n"
+            + "for person_id in person_ids:\n"
+            + "    person_id = person_id\n"
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:832]: Why?

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java [hunk 4]
```diff
@@ -815,6 +815,58 @@ private String getQualifiedColumnName(Domain currentDomain, String columnName) {
...
+            + "table = hl.import_table(f'{bucket}/"
+            + phenotypeFilename
+            + "', types={'sample_name': hl.tstr}, impute=True, key='sample_name')\n"
+            + "\n"
+            + "mt = hl.read_matrix_table(f'{bucket}/"
+            + cohortMtFilename
+            + "');\n"
+            + "mt = mt.annotate_cols(pheno = table[mt.s])\n"
+            + "\n"
+            + "gwas = hl.linear_regression_rows(y=mt.pheno.phenotype1, x=mt.GT.n_alt_alleles(), covariates=[1.0])\n"
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:865]: If covariates is something the user should expect to be modifying, then please comment or pull out into a variable

**Review by calbach — COMMENTED** [review]
Tests?

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java (hunk 1 above)
Comment (ericsong) [reply api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java]: The Hail tutorial had it as ".txt" but tsv makes sense.

**Comment by ericsong** [comment]
Should be good for second pass @calbach

**Review by calbach — APPROVED** [review]
Thanks for the tests, lgtm

File: api/src/test/java/org/pmiops/workbench/api/DataSetControllerTest.java [hunk 5]
```diff
@@ -739,4 +732,169 @@ public void testGetValuesFromDomain() {
...
+    assertThat(codeCells.get(4)).contains("plink");
+  }
+
+  List<String> notebookContentsToStrings(JSONObject notebookContents) {
+    List<String> codeCellStrings = new ArrayList<>();
+
+    JSONArray cells = notebookContents.getJSONArray("cells");
+    for (int i = 0; i < cells.length(); i++) {
+      String cellString = "";
+      JSONArray innerCells = cells.getJSONObject(i).getJSONArray("source");
```
[the remark this answers is not in the document]
Comment (ericsong) [reply api/src/test/java/org/pmiops/workbench/api/DataSetControllerTest.java:826]: I tried using streams but the issue I ran into is that the JSONArray type isn't iterable and it doesn't easily translate to streams. Would flatmap fix that issue?

[the remark this answers is not in the document]
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/api/DataSetControllerTest.java:826]: Ah I see. Probably not. You'd probably need to do something like:

```
.flatMap(StreamSupport.stream(
          Spliterators.spliteratorUnknownSize(innerCells.iterator()));
```

which is already getting pretty hairy. I'd leave it as you have it, in this case

**Pull request merged**